lib/PrintAsObjC/PrintAsObjC.cpp

```diff
@@ -803,7 +810,33 @@ class ObjCPrinter : private DeclVisitor<ObjCPrinter>,
       }
     }
   }
-    
+
+  void printSwift3ObjCDeprecatedInference(ValueDecl *VD) {
+    auto attr = VD->getAttrs().getAttribute<ObjCAttr>();
+    if (!attr || !attr->isSwift3Inferred())
```

Comment:
It'll end up tracking the language mode in which the declaration was compiled; we only mark `ObjCAttr` as "Swift 3 inferred" when Swift 3's `@objc` inference rules are applied.